Mathlib/Algebra/Category/ModuleCat/OverFilteredColimits.lean

```diff
@@ -0,0 +1,351 @@
+
+  apply_fun ℳ.map (IsFiltered.toSup O H (by simp [O])) at hn
+  rw [← comp_apply, ← comp_apply, ← ℳ.map_comp, ← ℳ.map_comp] at hn
+
+  rw [← comp_apply, ← comp_apply, ← comp_apply, ← comp_apply, ← ℛ.map_comp, ← ℛ.map_comp,
+    ← ℳ.map_comp, ← ℳ.map_comp]
+  convert congr($hm • $hn) using 1 <;> congr 3
+  · erw [IsFiltered.toSup_commutes O H (f := i₁), IsFiltered.toSup_commutes O H (f := fm₁)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := i₂), IsFiltered.toSup_commutes O H (f := fn₁)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := j₁), IsFiltered.toSup_commutes O H (f := fm₂)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := j₂), IsFiltered.toSup_commutes O H (f := fn₂)] <;>
+    simp [O, H]
+
+end hsmul
+
+variable {ℛ ℳ} in
+/--
+Let `R` be the filtered colimit of rings `{Rⱼ}` and `M` be the filtered colimit of
+abelian groups `{Mⱼ}`  with the same indexing set `j ∈ J`, if for each `j ∈ J`, `Mⱼ` is an `Rⱼ` such
+that the `Rⱼ`-action is compatible, then there is a scalar multiplication
+`Rⱼ → M → M` for every `j ∈ J`.
+-/
+noncomputable def smulColimit {c : J} (r : ℛ.obj c) (m : colimit (C := 𝔄𝔟) ℳ) :
+    colimit (C := 𝔄𝔟) ℳ :=
+  colimit.ι ℳ (IsFiltered.max c (Concrete.indexRepColimit ℳ m))
+   (hsmul (IsFiltered.leftToMax _ _) (IsFiltered.rightToMax _ _)
+    r (Concrete.repColimit ℳ m))
+
+namespace smulColimit
+
+lemma smul_rep (c₁ c₂ : J) (r : ℛ.obj c₁) (m : ℳ.obj c₂) :
+    smulColimit r (colimit.ι ℳ c₂ m) =
+    colimit.ι ℳ (IsFiltered.max c₁ c₂)
+    (hsmul (IsFiltered.leftToMax _ _) (IsFiltered.rightToMax _ _) r m) := by
+  delta smulColimit
+  apply hsmul.respect_ι
+  · rfl
+  · erw [Concrete.ι_repColimit_eq]
+
+protected lemma one_smul (c : J) (m : colimit (C := 𝔄𝔟) ℳ) :
+    smulColimit (1 : ℛ.obj c) m = m := by
+  rw [show m = colimit.ι ℳ (Concrete.indexRepColimit ℳ m) _ by
+    erw [Concrete.ι_repColimit_eq], smul_rep, hsmul.one_smul]
+  erw [colimit.w_apply]
+
+protected lemma mul_smul (c : J) (r₁ r₂ : ℛ.obj c)
+    (m : colimit (C := 𝔄𝔟) ℳ) :
+    smulColimit (r₁ * r₂) m = smulColimit r₁ (smulColimit r₂ m) := by
+  rw [show m = colimit.ι ℳ (Concrete.indexRepColimit ℳ m) _ by
+    erw [Concrete.ι_repColimit_eq], smul_rep, hsmul.mul_smul, smul_rep, smul_rep]
+  apply hsmul.respect_ι
+  · rfl
+  · apply hsmul.respect_ι
+    · rfl
+    · erw [Concrete.ι_repColimit_eq]
+
+lemma smul_zero (c : J) (r : ℛ.obj c) : smulColimit (ℳ := ℳ) r 0 = 0 := by
+  rw [show (0 : colimit (C := 𝔄𝔟) ℳ) = colimit.ι (C := 𝔄𝔟) ℳ c 0 by rw [map_zero],
+    smul_rep, hsmul.smul_zero, map_zero, map_zero]
+
+lemma smul_add (c : J) (r : ℛ.obj c) (m₁ m₂ : colimit (C := 𝔄𝔟) ℳ) :
+    smulColimit r (m₁ + m₂) = smulColimit r m₁ + smulColimit r m₂ := by
+  classical
+  let O : Finset J :=
+    { c, Concrete.indexRepColimit ℳ m₁, Concrete.indexRepColimit ℳ m₂ }
+  let H : Finset ((X : J) ×' (Y : J) ×' (_ : X ∈ O) ×' (_ : Y ∈ O) ×' (X ⟶ Y)) := {}
```

Comment:
```suggestion
```
You can remove it and then just replace all occurrences of `H` with `{}` if you like.